physics: Unit dyads access from ReferenceFrames

kostas-rigan opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
fixes #24965



#### Brief description of what is fixed or changed
Unit dyads and unit dyadic can be accessed through attributes of a ReferenceFrame object. It is an addition related to issue #24965, allowing easy access to unit dyads without the need to use the outer() function. This commit adds 10 functions with the property decorator, along with a few examples in the constructor's documentation.


#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->

* physics.vector
    * Added access to unit dyads and unit dyadic directly from ReferenceFrame objects.

<!-- END RELEASE NOTES -->

kostas-rigan committed:
physics: Unit dyads access from ReferenceFrames

Unit dyads can be accessed through attributes of a ReferenceFrame
object. It is an addition related to issue #24965, allowing easy access
to unit dyads without the need to use the outer() function. This commit
adds 9 functions with the @property decorator, along with a couple of
examples in the constructor. Access to unit dyadic is not implemented.

kostas-rigan committed:
Complied to code quality rules

moorepants commented:
This is looking good so far. You have some trailing whitespaces that the linter doesn't like:

```
sympy/testing/tests/test_code_quality.py:236: in test_this_file
    assert False, message_space % (fname, idx + 1)
E   AssertionError: File contains trailing whitespace: /home/runner/work/sympy/sympy/sympy/physics/vector/frame.py, line 1403.
E   assert False
```

moorepants commented:
The only think this needs are some simple unit tests to check the new functionality. Can you add a basic test for each of the new attributes?

kostas-rigan commented:
Yes I will write some tests. Sorry, I forgot to run the tests previously and didn't see the trailing whitespaces. 
I've also worked on adding the unit dyadic, and it looks like this:
``` python
@property
def dyadic(self):
    """Unit dyadic for the ReferenceFrame."""
    return Matrix([[1, 0, 0],
                   [0, 1, 0],
                   [0, 0, 1]])
```
I will write a test for it as well and then push everything together.

moorepants commented:
The unit dyadic should return a dyadic not a matrix.

It needs a better name too. We can do `ReferenceFrame.unit_dyadic` as that is the clearest. I was trying to think about a simpler name though because you may want to use it in math equations. Take the example here https://moorepants.github.io/learn-multibody-dynamics/mass.html#parallel-axis-theorem

```python
dx, dy, dz, m = sm.symbols('d_x, d_y, d_z, m')

N = me.ReferenceFrame('N')

r_O_Bo = dx*N.x + dy*N.y + dz*N.z

U = me.outer(N.x, N.x) + me.outer(N.y, N.y) + me.outer(N.z, N.z)

I_Bo_O = m*(me.dot(r_O_Bo, r_O_Bo)*U - me.outer(r_O_Bo, r_O_Bo))
```

it could be written like:

```python
I_Bo_O = m*(me.dot(r_O_Bo, r_O_Bo)*N.u - me.outer(r_O_Bo, r_O_Bo))
```

or something. Much like our `N.x,N.xx,`.

moorepants commented:
I guess `N.unit` could work. It is shorter and maybe descriptive enough. I'll think about the best name a bit.

@TJStienstra any preferences on a name for a unit dyadic?

abhishek-kuma commented:
This is my first time contributing to an open source, so I'm less experienced. The issue is almost solved. I'm actually contributing for GSoC - 2023 . I think i have to learn how to add tests now .

moorepants commented:
>  I think i have to learn how to add tests now .

Yes, all PRs should have some unit tests.

kostas-rigan commented:
> The unit dyadic should return a dyadic not a matrix.
> 
I ran your example, and I changed the code to:
``` python
@property
def unit(self):
    """Unit dyadic for the ReferenceFrame."""
    return self.xx + self.yy + self.zz
```

kostas-rigan committed:
physics: added unit dyadic

Added unit dyadic as attribute u. Tests were also implemented for both unit
dyads and unit dyadic.

kostas-rigan committed:
Fixed unit dyad example documentation mistake

kostas-rigan commented:
@moorepants is it preferable to squash the commits into a single one? Due to some mistakes I made, there are a couple of commits for minor changes that I could have avoided.

moorepants commented:
> @moorepants is it preferable to squash the commits into a single one? Due to some mistakes I made, there are a couple of commits for minor changes that I could have avoided.

Squashing is not necessary or very common in SymPy commit history, but if you know what you are doing and want to force push a cleaned up commit history you can.

moorepants commented on sympy/physics/vector/frame.py:
```diff
@@ -145,6 +145,30 @@ def __init__(self, name, indices=None, latexs=None, variables=None):
+
+        >>> from sympy.physics.vector import ReferenceFrame
+        >>> N = ReferenceFrame('N')
+        >>> N.unit
```
```suggestion
        >>> N.u
```

moorepants commented:
If you fix the minor error in the doctest and then update the release notes to reflect adding unit dyads and the unit dyadic, then this looks good to me.

kostas-rigan commented:
> If you fix the minor error in the doctest and then update the release notes to reflect adding unit dyads and the unit dyadic, then this looks good to me.

Great, I will fix that right away. Thank you for the guidance!

kostas-rigan committed:
fixed doctest error in unit dyadic example

kostas-rigan commented:
> #### References to other Issues or PRs
> fixes #24965
> 
> #### Brief description of what is fixed or changed
> Unit dyads and unit dyadic can be accessed through attributes of a ReferenceFrame object. It is an addition related to issue #24965, allowing easy access to unit dyads without the need to use the outer() function. This commit adds 10 functions with the property decorator, along with a few of examples in the constructor documentation.
> 
> #### Other comments
> 
> 
> #### Release Notes
> * physics.vector
>   * Added access to unit dyads and unit dyadic directly from ReferenceFrame objects.

moorepants commented:
Thanks for contribution. If tests pass it will merge!
